dbms/src/Functions/runningAccumulate.cpp

```diff
@@ -70,33 +69,53 @@ class FunctionRunningAccumulate : public IFunction
 
     void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t /*input_rows_count*/) override
     {
+        size_t number_of_arguments = arguments.size();
+
         const ColumnAggregateFunction * column_with_states
             = typeid_cast<const ColumnAggregateFunction *>(&*block.getByPosition(arguments.at(0)).column);
+
         if (!column_with_states)
             throw Exception("Illegal column " + block.getByPosition(arguments.at(0)).column->getName()
                     + " of first argument of function "
                     + getName(),
                 ErrorCodes::ILLEGAL_COLUMN);
 
+        ColumnPtr column_with_groups;
+
+        if (number_of_arguments == 2)
+            column_with_groups = block.getByPosition(arguments[1]).column;
+
         AggregateFunctionPtr aggregate_function_ptr = column_with_states->getAggregateFunction();
         const IAggregateFunction & agg_func = *aggregate_function_ptr;
 
         AlignedBuffer place(agg_func.sizeOfData(), agg_func.alignOfData());
         agg_func.create(place.data());
         SCOPE_EXIT(agg_func.destroy(place.data()));
 
+        /// Will pass empty arena if agg_func does not allocate memory in arena
         std::unique_ptr<Arena> arena = agg_func.allocatesMemoryInArena() ? std::make_unique<Arena>() : nullptr;
 
         auto result_column_ptr = agg_func.getReturnType()->createColumn();
         IColumn & result_column = *result_column_ptr;
         result_column.reserve(column_with_states->size());
 
         const auto & states = column_with_states->getData();
+
+        size_t i = 0;
+
         for (const auto & state_to_add : states)
         {
-            /// Will pass empty arena if agg_func does not allocate memory in arena
+            if (column_with_groups && i > 0 &&
+                column_with_groups->compareAt(i, i - 1, *column_with_groups, 1) != 0)
+            {
+                agg_func.destroy(place.data());
+                agg_func.create(place.data());
```

Comment:
Need something for exception safety if create will fail.
For example, you can add a bool flag and check it in SCOPE_EXIT above.

Reply:
> `SELECT runningAccumulate()` leads to crash.

Now arguments count are tested before invocation.

Reply:
There is `SCOPE_EXIT` with `agg_func.destroy` callback installed at [line 93](https://github.com/ClickHouse/ClickHouse/pull/8326/files/c339c1ca665de7f87322a2522c5753fecacd8839#diff-b38a336c2370592dbcf9c063a1e2927cR93). And if I understood everything correctly
1. `SCOPE_EXIT` uses destructor and it will be destroyed in case of exception too;
2. `agg_func.destroy` (as I saw in other places) must be called, whenever creation failure occurred.

Is it really need to handle `agg_func.create` calling exceptions?

Reply:
Yes, create may fail due to memory limits. In that case, destroy will be called twice. And it may lead to double free.

Reply:
Done!